presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -454,6 +494,48 @@ public HashAggregationBuilder getAggregationBuilder()
         return aggregationBuilder;
     }
 
+    private void processInputPage(Page page)
+    {
+        if (preGroupedHashStrategy.isPresent()) {
+            // Running in segmented aggregation mode.
+            int lastRowInPage = page.getPositionCount() - 1;
+            if (firstUnfinishedSegment == null) {
+                // If this is the first page, treat the first segment in this page as the current segment.
+                firstUnfinishedSegment = page.getRegion(0, 1);
+            }
+
+            if (preGroupedHashStrategy.get().rowEqualsRow(0, firstUnfinishedSegment.extractChannels(preGroupedChannels), lastRowInPage, page.extractChannels(preGroupedChannels))) {
+                // If the current segment might have more data in the incoming pages, process the whole page.
+                unfinishedWork = aggregationBuilder.processPage(page);
+            }
+            else if (preGroupedHashStrategy.get().rowEqualsRow(0, firstUnfinishedSegment.extractChannels(preGroupedChannels), 0, page.extractChannels(preGroupedChannels))) {
```

Comment:
If this is the first page, line 504 is executed, no need to make this comparison. Please try to avoid it in that case.